src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,977 @@
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                var memberDeclarationNode = FindParentMemberDeclarationNode(returnStatement);
+                if (memberDeclarationNode == null)
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                var declaredSymbol = _semanticModel.GetDeclaredSymbol(memberDeclarationNode);
+                ITypeSymbol typeSymbol;
+                switch (declaredSymbol.Kind)
+                {
+                    case SymbolKind.Local:
+                        typeSymbol = ((ILocalSymbol)declaredSymbol).Type;
+                        break;
+                    case SymbolKind.Method:
+                        typeSymbol = ((IMethodSymbol)declaredSymbol).ReturnType;
+                        break;
+                    case SymbolKind.Property:
+                        typeSymbol = ((IPropertySymbol)declaredSymbol).Type;
+                        break;
+                    default:
+                        throw new ArgumentException(declaredSymbol.Kind.ToString());
+                }
+
+                if (typeSymbol.OriginalDefinition?.SpecialType != SpecialType.System_Collections_Generic_IEnumerable_T)
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // if there are more than one return in the method, convert to local funciton.
+                if (memberDeclarationNode.DescendantNodes().OfType<ReturnStatementSyntax>().Count() == 1)
+                {
+                    // before: return from a in b select a;
+                    // after: 
+                    // foreach(var a in b)
+                    // {
+                    //      yield return a;
+                    // }
+                    //
+                    // yield break;
+                    var statements = GenerateStatements((ExpressionSyntax expression)
+                        => SyntaxFactory.YieldStatement(SyntaxKind.YieldReturnStatement, expression), queryExpressionProcessingInfo);
+
+                    // add an yield break to avoid throws after the return.
+                    var yieldBreakStatement = SyntaxFactory.YieldStatement(SyntaxKind.YieldBreakStatement);
+                    documentUpdateInfo = new DocumentUpdateInfo(returnStatement, statements.Concat(new[] { yieldBreakStatement }));
+                    return true;
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            // We may assume that the query is defined within a method, field, property and so on and it is declare just once.
+            private SyntaxNode FindParentMemberDeclarationNode(SyntaxNode node)
+                => _semanticModel.GetEnclosingSymbol(node.SpanStart, _cancellationToken).DeclaringSyntaxReferences.First().GetSyntax();
```

Comment:
>First() [](start = 115, length = 7)

Is this safe on partial methods? For example, how many DeclaringSyntaxReferences returned for partial method implementation? If we are sure there is always only one, we should use Single, otherwise we should probably pick the one that contains the query (span check should be sufficient).

Reply:
Added a test with partial methods. DeclaringSyntaxReferences provides a single one.

---
In reply to: [178423361](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423361) [](ancestors = 178423361)

Reply:
Let us use Single

---
In reply to: [178423710](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423710) [](ancestors = 178423710,178423361)